tests/Module/RateCalculatorTest.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+declare( strict_types=1 );
+
+namespace Tests\Module;
+
+use Packetery\Module\Carrier\Options;
+use Packetery\Module\RateCalculator;
+use PHPUnit\Framework\TestCase;
+
+class RateCalculatorTest extends TestCase {
+	use WithMockFactory;
+
+	/**
+	 * @return array
+	 */
+	public static function calculationDataProvider(): array {
+		$defaultCarrierOptionsArray = [
+			'id'                   => '106',
+			'free_shipping_limit'  => 10000.0,
+			'pricing_type'         => Options::PRICING_TYPE_BY_WEIGHT,
+			'weight_limits'        => [
+				[
+					'weight' => 10,
+					'price'  => 11,
+				],
+				[
+					'weight' => 20,
+					'price'  => 22,
+				],
+				[
+					'weight' => 30,
+					'price'  => 33,
+				],
+			],
+			'product_value_limits' => [
+				[
+					'value' => 100,
+					'price' => 111,
+				],
+				[
+					'value' => 200,
+					'price' => 222,
+				],
+				[
+					'value' => 300,
+					'price' => 333,
+				],
+			],
+			'coupon_free_shipping' => [
+				'active'         => true,
+				'allow_for_fees' => false,
+			],
+		];
+
+		$defaultCarrierOptions             = new Options( 'any', $defaultCarrierOptionsArray );
+		$carrierOptionsNoCoupon            = new Options(
+			'any',
+			array_merge(
+				$defaultCarrierOptionsArray,
+				[
+					'coupon_free_shipping' => [
+						'active' => false,
+					],
+				]
+			)
+		);
+		$carrierOptionsProductValuePricing = new Options(
+			'any',
+			array_merge(
+				$defaultCarrierOptionsArray,
+				[
+					'pricing_type' => Options::PRICING_TYPE_BY_PRODUCT_VALUE,
+				]
+			)
+		);
+		$carrierOptionsNoFreeShippingLimit = new Options(
+			'any',
+			array_merge(
+				$defaultCarrierOptionsArray,
+				[
+					'free_shipping_limit' => null,
+				]
+			)
+		);
+
+		return [
+			[ 11, $defaultCarrierOptions, 100, 0, false ],
+			[ 22, $defaultCarrierOptions, 100, 15, false ],
+			[ 22, $defaultCarrierOptions, 100, 20, false ],
+			[ 33, $defaultCarrierOptions, 100, 25, false ],
+			[ null, $defaultCarrierOptions, 100, 31, false ],
+			[ 111, $carrierOptionsProductValuePricing, 100, 1, false ],
+			[ 222, $carrierOptionsProductValuePricing, 150, 15, false ],
+			[ 222, $carrierOptionsProductValuePricing, 200, 20, false ],
+			[ 333, $carrierOptionsProductValuePricing, 300, 25, false ],
+			[ null, $carrierOptionsProductValuePricing, 400, 31, false ],
+			'free shipping threshold must be reached'     => [
+				0,
+				$defaultCarrierOptions,
+				15000,
```

Comment:
Consider to use array keys to improving readability - I don't know what is 15000 value, 25 and false.
`'totalProductValue' => 15000`